src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -35,21 +36,27 @@ internal abstract class LanguageServerProjectLoader
     private readonly ProjectTargetFrameworkManager _targetFrameworkManager;
     private readonly ProjectSystemHostInfo _projectSystemHostInfo;
     private readonly IFileChangeWatcher _fileChangeWatcher;
-    private readonly IGlobalOptionService _globalOptionService;
+    protected readonly IGlobalOptionService GlobalOptionService;
     protected readonly ILoggerFactory LoggerFactory;
     private readonly ILogger _logger;
     private readonly ProjectLoadTelemetryReporter _projectLoadTelemetryReporter;
     private readonly BinlogNamer _binlogNamer;
-    private readonly ProjectFileExtensionRegistry _projectFileExtensionRegistry;
     protected readonly ImmutableDictionary<string, string> AdditionalProperties;
 
     /// <summary>
-    /// The list of loaded projects in the workspace, keyed by project file path. The outer dictionary is a concurrent dictionary since we may be loading
-    /// multiple projects at once; the key is a single List we just have a single thread processing any given project file. This is only to be used
-    /// in <see cref="LoadOrReloadProjectsAsync" /> and downstream calls; any other updating of this (like unloading projects) should be achieved by adding
-    /// things to the <see cref="ProjectsToLoadAndReload" />.
+    /// Stores LoadedProjects representing all the TFMs of a single project (whose file path is the key in <see cref="_loadedProjects"/>)
     /// </summary>
-    private readonly ConcurrentDictionary<string, List<LoadedProject>> _loadedProjects = [];
+    /// <param name="Semaphore">
+    /// Synchronizes access to and disposal of items in <see cref="LoadedProjects"/>.
+    /// This must be obtained prior to any actions on LoadedProjects.
+    /// </param>
+    /// <param name="CancellationTokenSource">
+    /// When cancelled, signals that some thread has started disposing the <see cref="LoadedProjects"/>.
+    /// This must be checked after acquiring <see cref="Semaphore"/>, prior to performing any actions on LoadedProjects.
+    /// </param>
+    protected record LoadedProjectSet(List<LoadedProject> LoadedProjects, SemaphoreSlim Semaphore, CancellationTokenSource CancellationTokenSource);
+
+    private readonly ConcurrentDictionary<string, LoadedProjectSet> _loadedProjects = [];
```

Comment:
1. what is the key here?
2. why use a ConDict, if a semaphore is needed to access the data within.  Do we really need fine grained concurrency accesso that level?